src/main/java/io/aiven/kafka/connect/common/grouper/TopicPartitionRecordGrouper.java

```diff
@@ -0,0 +1,151 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package io.aiven.kafka.connect.common.grouper;
+
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.connect.sink.SinkRecord;
+
+import io.aiven.kafka.connect.common.config.FilenameTemplateVariable;
+import io.aiven.kafka.connect.common.config.TimestampSource;
+import io.aiven.kafka.connect.common.templating.Template;
+import io.aiven.kafka.connect.common.templating.VariableTemplatePart.Parameter;
+
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * A {@link RecordGrouper} that groups records by topic and partition.
+ *
+ * <p>The class requires a filename template with {@code topic}, {@code partition},
+ * and {@code start_offset} variables declared.
+ *
+ * <p>The class supports limited and unlimited number of records in files.
+ */
+public final class TopicPartitionRecordGrouper implements RecordGrouper {
+
+    private final Template filenameTemplate;
+
+    private final Integer maxRecordsPerFile;
+
+    private final Map<TopicPartition, SinkRecord> currentHeadRecords = new HashMap<>();
+
+    private final Map<String, List<SinkRecord>> fileBuffers = new HashMap<>();
+
+    private final Function<Parameter, String> setTimestamp;
+
+    /**
+     * A constructor.
+     *
+     * @param filenameTemplate  the filename template.
+     * @param maxRecordsPerFile the maximum number of records per file ({@code null} for unlimited).
+     */
+    public TopicPartitionRecordGrouper(final Template filenameTemplate,
+                                       final Integer maxRecordsPerFile,
+                                       final TimestampSource tsSource) {
+        Objects.requireNonNull(filenameTemplate, "filenameTemplate cannot be null");
+        Objects.requireNonNull(tsSource, "tsSource cannot be null");
+        this.filenameTemplate = filenameTemplate;
+        this.maxRecordsPerFile = maxRecordsPerFile;
+        this.setTimestamp = new Function<Parameter, String>() {
+
+            //FIXME move into commons lib
+            private final Map<String, DateTimeFormatter> timestampFormatters =
```

Comment:
This function was defined in `FomratUtils` lets use it

Reply:
It is already prepared in other PR